aws-s3-transfer-manager/src/operation/upload/checksum_strategy.rs

```diff
@@ -0,0 +1,283 @@
+            full_object_checksum: None,
+        }
+    }
+
+    /// The transfer manager calculates a `CRC32C` full object checksum while uploading.
+    pub fn with_calculated_crc32_c() -> Self {
+        Self {
+            algorithm: ChecksumAlgorithm::Crc32C,
+            type_if_multipart: ChecksumType::FullObject,
+            full_object_checksum: None,
+        }
+    }
+
+    /// The transfer manager calculates `CRC32` checksums while uploading.
+    /// If the upload is multipart, the object will have a composite checksum,
+    /// otherwise it will have a full object checksum.
+    pub fn with_calculated_crc32_composite_if_multipart() -> Self {
+        Self {
+            algorithm: ChecksumAlgorithm::Crc32,
+            type_if_multipart: ChecksumType::Composite,
+            full_object_checksum: None,
+        }
+    }
+
+    /// The transfer manager calculates `CRC32C` checksums while uploading.
+    /// If the upload is multipart, the object will have a composite checksum,
+    /// otherwise it will have a full object checksum.
+    pub fn with_calculated_crc32_c_composite_if_multipart() -> Self {
+        Self {
+            algorithm: ChecksumAlgorithm::Crc32C,
+            type_if_multipart: ChecksumType::Composite,
+            full_object_checksum: None,
+        }
+    }
+
+    /// The transfer manager calculates `SHA1` checksums while uploading.
+    /// If the upload is multipart, the object will have a composite checksum,
+    /// otherwise it will have a full object checksum.
+    pub fn with_calculated_sha1_composite_if_multipart() -> Self {
+        Self {
+            algorithm: ChecksumAlgorithm::Sha1,
+            type_if_multipart: ChecksumType::Composite,
+            full_object_checksum: None,
+        }
+    }
+
+    /// The transfer manager calculates `SHA256` checksums while uploading.
+    /// If the upload is multipart, the object will have a composite checksum,
+    /// otherwise it will have a full object checksum.
+    pub fn with_calculated_sha256_composite_if_multipart() -> Self {
+        Self {
+            algorithm: ChecksumAlgorithm::Sha256,
+            type_if_multipart: ChecksumType::Composite,
+            full_object_checksum: None,
+        }
+    }
+
+    /// Create checksum strategy from parts.
+    ///
+    /// It is possible to create an invalid strategy this way,
+    /// use [`ChecksumStrategy::validate()`] to check. e.g. S3 does not allow
+    /// any `SHA` algorithm to do full object multipart uploads, and does not
+    /// allow `CRC64NVME` to do composite multipart uploads.
+    pub fn new(
+        algorithm: ChecksumAlgorithm,
+        type_if_multipart: ChecksumType,
+        full_object_checksum: Option<String>,
+    ) -> Self {
+        Self {
+            algorithm,
+            type_if_multipart,
+            full_object_checksum,
+        }
+    }
+
+    pub(crate) fn validate(&self) -> Result<(), crate::error::Error> {
```

Comment:
PR description mentions this is public not just pub crate.

Reply:
this function is gone now, `Builder::build()` does validation